conmon-rs/server/src/attach.rs

```diff
@@ -46,43 +59,134 @@ impl Attach {
+                            .expect("unable to write stderr");
+                    }
+                }
+            }
+        });
         Ok(())
     }
 
-    fn handle_client(stream: UnixStream) -> Result<()> {
-        debug!("Got new attach stream connection");
+    /// Try to read from all streams standard input and return the first result.
+    pub async fn try_read(&mut self) -> Result<Vec<u8>> {
+        let mut cleanup_idxs = vec![];
+        let mut clients = self.clients.write().await;
 
-        let mut reader = BufReader::new(&stream);
-        let mut writer = BufWriter::new(&stream);
+        for (idx, stream) in clients.iter().enumerate() {
+            let ready = stream.ready(Interest::READABLE).await?;
 
-        const BUF_SIZE: usize = 8192;
-        let mut buf = vec![0; BUF_SIZE];
+            // Cleanup closed channels
+            if ready.is_read_closed() {
+                debug!("Cleanup stale client index: {}", idx);
+                cleanup_idxs.push(idx);
+                continue;
+            }
 
-        // TODO: implement me
-        buf[0] = 2;
-        writer.write_all(&buf)?;
-        buf[0] = 3;
-        writer.write_all(&buf)?;
+            if ready.is_readable() {
+                let mut buf = vec![0; ATTACH_PACKET_BUF_SIZE];
+                match stream.try_read(&mut buf) {
+                    Ok(n) if n > 0 => {
+                        if let Some(first_zero_idx) = buf.iter().position(|&x| x == 0) {
+                            buf.resize(first_zero_idx, 0);
+                        }
+                        debug!("Read {} stdin bytes from client", buf.len());
+                        Self::cleanup_clients(&mut clients, &cleanup_idxs).await;
+                        return Ok(buf);
+                    }
+                    Err(ref e) if e.kind() == ErrorKind::WouldBlock => continue,
+                    Err(e) => {
+                        Self::cleanup_clients(&mut clients, &cleanup_idxs).await;
+                        return Err(e.into());
+                    }
+                    _ => {}
+                }
+            }
+        }
 
-        loop {
-            match reader.read(&mut buf) {
-                Ok(n) if n > 0 => {
-                    debug!("Read {} bytes", n);
+        Self::cleanup_clients(&mut clients, &cleanup_idxs).await;
+
+        // Nothing to read, return an empty result.
+        Ok(vec![])
+    }
+
+    async fn cleanup_clients(clients: &mut Vec<UnixStream>, idxs: &[usize]) {
+        for i in idxs.iter().rev() {
+            clients.remove(*i);
+        }
+    }
+
+    /// Write a buffer to all attached clients.
+    pub async fn write<T>(&self, pipe: Pipe, buf: T) -> Result<()>
+    where
+        T: AsRef<[u8]>,
+    {
+        let packets = buf
+            .as_ref()
+            .chunks(ATTACH_PACKET_BUF_SIZE - 1)
+            .map(|x| {
+                let mut y = x.to_vec();
+                let p = match pipe {
+                    Pipe::StdOut => 2,
```

Comment:
we should have a TODO here to not go over stdout/stderr but the container's pipes

Reply:
given how this works, should we save the container's fds in the container_io instance and save that in the child instance?

Reply:
I think in long term we have to store the container IOs (the one from createcontainer and all execs) in the child and reference them by IDs. This way we can store the attach inside of the corresponding container IO on the attach RPC.

The `try_read` and `write` functions will be then called by the global read loop.